### PR TITLE
feat: Do not merge version when live version is more recent

### PR DESCRIPTION
**Pull request by jasperP98**
### 1. Why is this change necessary?
When you have a version from the entity and in the meantime the live version of that entity is updated, and then you merge the old version, the live version its change is overwritten by the old version. We should block this merge.

### 2. What does this change do, exactly?
This change will block merges when a version is older than the latest update from its live version.

### 3. Describe each step to reproduce the issue or behaviour.
1. Open an order in the admin
2. Do a change to that order (like adding a product) but do not save
3. Open the same order in another window
4. Set the order state to completed and save it
5. Go to the other window and save
6. Now the state will be back to open

### 5. Checklist

- [ ] I have written tests and verified that they fail without my change
- [X] I have created a [changelog file](https://github.com/shopware/shopware/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [X] I have read the contribution requirements and fulfilled them

### CODE REVIEW

File: changelog/_unreleased/2025-06-27-improve-version-timestamp-checking.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
+---
+title: Improve version timestamp checking
+issue: NEXT-00000
```
Comment (mitelg) [inline changelog/_unreleased/2025-06-27-improve-version-timestamp-checking.md:3]:
```suggestion
```

File: changelog/_unreleased/2025-06-27-improve-version-timestamp-checking.md [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+author: Jasper Peeters
+author_email: jasper.peeters@meteor.be
+author_github: JasperP98
+---
+
+# Core
+* Changed: Added a merge block (`Shopware\Core\Framework\DataAbstractionLayer\VersionManager::checkVersionTimestamps`) when a version is trying to merge that is older than the live version
```
Comment (mitelg) [inline changelog/_unreleased/2025-06-27-improve-version-timestamp-checking.md:10]:
```suggestion
* Added a merge block (`Shopware\Core\Framework\DataAbstractionLayer\VersionManager::checkVersionTimestamps`) when a version is trying to be merged that is older than the live version
```

File: src/Core/Framework/DataAbstractionLayer/DataAbstractionLayerException.php [hunk 3]
```diff
@@ -235,6 +236,16 @@ public static function versionNotExists(string $versionId): self
         );
     }
 
+    public static function versionIsOlderThanLiveVersion(string $versionId): self
+    {
+        return new self(
+            Response::HTTP_BAD_REQUEST,
+            self::VERSION_IS_OLDER_THAN_LIVE_VERSION,
+            'Version {{ versionId }} is older than the live version. Cannot merge.',
```
Comment (mitelg) [inline src/Core/Framework/DataAbstractionLayer/DataAbstractionLayerException.php:244]:
```suggestion
            'Version "{{ versionId }}" cannot be merged, as it is older than the live version.',
```

File: src/Core/Framework/DataAbstractionLayer/VersionManager.php [hunk 4]
```diff
@@ -2,6 +2,7 @@
 
 namespace Shopware\Core\Framework\DataAbstractionLayer;
 
+use LensOnline\Exception\VersionMergeOutdatedException;
```
Comment (mitelg) [inline src/Core/Framework/DataAbstractionLayer/VersionManager.php:5]:
```suggestion
```

File: src/Core/Framework/DataAbstractionLayer/VersionManager.php [hunk 5]
```diff
@@ -168,6 +169,8 @@ public function merge(string $versionId, WriteContext $writeContext): void
             throw DataAbstractionLayerException::versionNotExists($versionId);
         }
 
+        $this->checkVersionTimestamps($versionId, $writeContext);
```
Comment (mitelg) [inline src/Core/Framework/DataAbstractionLayer/VersionManager.php:172]: within this method you are also calling `$this->getCommits`. wouldn't it be better, if you call that before and pass the commits into the new method? so there is no duplicate DB call

File: src/Core/Framework/DataAbstractionLayer/VersionManager.php [hunk 6]
```diff
@@ -212,6 +215,44 @@ public function merge(string $versionId, WriteContext $writeContext): void
...
     }
 
+    private function checkVersionTimestamps(string $versionId, WriteContext $writeContext): void
+    {
+        $commits = $this->getCommits($versionId, $writeContext);
+        if ($commits->count() === 0) {
+            return;
+        }
+
+        foreach ($commits as $commit) {
```
Comment (mitelg) [inline src/Core/Framework/DataAbstractionLayer/VersionManager.php:225]: we are already iterating over the commits a lot. wouldn't it be possible to combine those?

File: src/Core/Framework/DataAbstractionLayer/VersionManager.php [hunk 7]
```diff
@@ -212,6 +215,44 @@ public function merge(string $versionId, WriteContext $writeContext): void
...
+                }
+
+                $entityName = $commitData->getEntityName();
+                $entityId = $commitData->getEntityId();
+
+                if (!$entityName || !$entityId) {
+                    continue;
+                }
+
+                $liveEntity = $this->entityReader->read(
```
Comment (mitelg) [inline src/Core/Framework/DataAbstractionLayer/VersionManager.php:239]: what is the performance impact of reading each entity first before we can check if a merge is possible? :thinking: 

we need to consider if this is worth it to only prevent a probable edge case while using the UI not correctly

**Review by mitelg — CHANGES_REQUESTED** [review]
would be nice, if you could add some tests

File: src/Core/Framework/DataAbstractionLayer/VersionManager.php (hunk 7 above)
Comment (shyim) [reply src/Core/Framework/DataAbstractionLayer/VersionManager.php:239]: can we do plain sql and do an early return when the entity does not have an updatedAt field?

we don't need to fetch the entire entity just to compare updatedAt. and you're doing the updatedAt in an for each `:n` query

**Comment by mitelg** [comment]
hey @jasperP98 

are you still working on that?

**Comment by mitelg** [comment]
Hey @jasperP98 

This PR was closed due to inactivity. If this change is still important to you, feel free to create a new pull request. 

For more information about our contribution guidelines, see https://github.com/shopware/shopware/blob/trunk/CONTRIBUTING.md
